streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java

```
@@ -0,0 +1,347 @@
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.serialization.ByteArraySerializer;
+import org.apache.kafka.common.serialization.BytesSerializer;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.internals.ProcessorRecordContext;
+import org.apache.kafka.streams.processor.internals.ProcessorStateManager;
+import org.apache.kafka.streams.processor.internals.RecordBatchingStateRestoreCallback;
+import org.apache.kafka.streams.processor.internals.RecordCollector;
+import org.apache.kafka.streams.state.StoreBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+public class InMemoryTimeOrderedKeyValueBuffer implements TimeOrderedKeyValueBuffer {
+    private static final Logger LOG = LoggerFactory.getLogger(InMemoryTimeOrderedKeyValueBuffer.class);
+    private static final BytesSerializer KEY_SERIALIZER = new BytesSerializer();
+    private static final ByteArraySerializer VALUE_SERIALIZER = new ByteArraySerializer();
+
+    private final Map<Bytes, BufferKey> index = new HashMap<>();
+    private final TreeMap<BufferKey, ContextualRecord> sortedMap = new TreeMap<>();
+
+    private final Set<Bytes> dirtyKeys = new HashSet<>();
+    private final String storeName;
+    private final boolean loggingEnabled;
+
+    private long memBufferSize = 0L;
+    private long minTimestamp = Long.MAX_VALUE;
+    private boolean open;
```

Do we expect `TimeOrderedKeyValueBuffer` to be queryable ever (I saw you changed it to extend StateStore)? If yes we need to make this variable as volatile since IQ threads may access `isOpen`.

---

Is `isOpen` used only by IQ? In that case, maybe we should throw an unsupported operation exception.

I expect the buffer never to be queriable. That would seem to be a very strange pattern.

---

I went ahead and made it volatile anyway, just in case.